crates/marketplace/src/testing/order_test.rs

```diff
@@ -0,0 +1,527 @@
+                prev_proposed_transactions = Some(proposed_transactions);
+            }
+        } else {
+            tracing::error!("Unable to get request from RequestMessage");
+        }
+        // save transactions to history
+        if prev_proposed_transactions.is_some() {
+            transaction_history.extend(prev_proposed_transactions.clone().unwrap());
+        }
+    }
+
+    // we should've served all transactions submitted, and in correct order
+    // the test will fail if the common part of two vectors of transactions don't have the same order
+    assert!(order_check(transaction_history, all_transactions));
+}
+
+/// This test simulates multiple builder states receiving messages from the channels and processing them
+/// and focus specifically on orders with chain fork.
+/// with one chain proposing transactions we've given and the other not
+/// (we should give out the next batch if responding for first chain and both batches for the other)
+#[async_std::test]
+async fn test_builder_order_chain_fork() {
+    async_compatibility_layer::logging::setup_logging();
+    async_compatibility_layer::logging::setup_backtrace();
+    tracing::info!("Testing the builder core with multiple messages from the channels");
+
+    // Number of views to simulate
+    const NUM_ROUNDS: usize = 3;
+    // Number of transactions to submit per round
+    const NUM_TXNS_PER_ROUND: usize = 5;
+    // Capacity of broadcast channels
+    const CHANNEL_CAPACITY: usize = NUM_ROUNDS * 5;
+    // Number of nodes on DA committee
+    const NUM_STORAGE_NODES: usize = 4;
+
+    let (senders, global_state) = start_builder_state(CHANNEL_CAPACITY, NUM_STORAGE_NODES).await;
+
+    // Transactions to send
+    let all_transactions = (0..NUM_ROUNDS)
+        .map(|round| {
+            (0..NUM_TXNS_PER_ROUND)
+                .map(|tx_num| TestTransaction::new(vec![round as u8, tx_num as u8]))
+                .collect::<Vec<_>>()
+        })
+        .collect::<Vec<_>>();
+
+    let fork_round = 1; // the round we want to skip all the transactions for the fork chain
+
+    // set up state to track between simulated consensus rounds
+    let mut prev_proposed_transactions: Option<Vec<TestTransaction>> = None;
+    let mut prev_quorum_proposal: Option<QuorumProposal<TestTypes>> = None;
+    let mut transaction_history = Vec::new();
+
+    // set up state to track the fork-ed chain
+    let mut prev_proposed_transactions_2: Option<Vec<TestTransaction>> = None;
+    let mut prev_quorum_proposal_2: Option<QuorumProposal<TestTypes>> = None;
+    let mut transaction_history_2 = Vec::new();
+    let mut fork: bool;
+
+    // Simulate NUM_ROUNDS of consensus. First we submit the transactions for this round to the builder,
+    // then construct DA and Quorum Proposals based on what we received from builder in the previous round
+    // and request a new bundle.
+    #[allow(clippy::needless_range_loop)] // intent is clearer this way
+    for round in 0..NUM_ROUNDS {
```

Comment:
Good suggestion! Now I set `NUM_ROUNDS` to 3 and `fork_round` to 1 so that we have round 1 as a round the fork happens and round 2 as a round that there's really a fork. But I'll update to what you suggested.

Reply:
When I'm doing the change I found that this will add more duplicate codes, also it's hard to track when the fork does not exist anymore. In practice it's possible that proposers keep falling behind and gradually catch up (although this is not expected in the test), therefore we don't know when would be after-fork phase. Now I just add more comments and use round also as `if` condition instead of simply using `fork`. Would this be helpful?